### PR TITLE
Implement meteor environment's packageDependencies

### PR DESCRIPTION
**Pull request by rhettlivingston**
Meteor includes two package systems, its own and support for NPM packages. Currently, only NPM packages are being recognized as such by the default packageDependencies implementation.

This submission builds on #303 which allows environment configurations to supply a packageDependencies function.

A packageDependencies function is added to environments/meteorEnvironment.js. It returns a list that is a combination of Meteor packages extracted from the .meteor/packages file and NPM packages found via the existing findPackageDependencies function.

### CODE REVIEW

**Comment by rhettlivingston** [comment]
Tests to follow. Wanted to get it out for comments.

File: lib/environments/meteorEnvironment.js [hunk 1]
```diff
@@ -1,22 +1,46 @@
...
+function meteorPackageDependencies(
+  { config }: Object
+): Array<string> {
+  const meteorPackagesPath = path.join(config.workingDirectory, '.meteor/packages');
+  if (!fs.existsSync(meteorPackagesPath)) {
+    return [];
+  }
+  const packages = fs.readFileSync(meteorPackagesPath, 'utf8')
+    // extract an array of package names from the packages file
+    .match(/^\s*([^# @\n]+)/gm)
```
Comment (trotzig) [inline lib/environments/meteorEnvironment.js]: I could be wrong, but I don't think the `m` flag exists for javascript regexps. We've been using https://www.npmjs.com/package/xregexp for that, but I'm not sure you need it here. 

Could packages ever have `@` in their name? It looks like we would ignore them in such case.

File: lib/environments/meteorEnvironment.js [hunk 2]
```diff
@@ -1,22 +1,46 @@
...
+  'meteor/reactive-var',
+  'meteor/session',
+  'meteor/templating',
+  'meteor/tracker',
+];
+
+function meteorPackageDependencies(
+  { config }: Object
+): Array<string> {
+  const meteorPackagesPath = path.join(config.workingDirectory, '.meteor/packages');
```
Comment (trotzig) [inline lib/environments/meteorEnvironment.js]: I'm trying to find documentation about what the syntax of this file looks like. Based on the regexp it seems to be a simple text file with a package per line. It could help if there was a comment here explaining things, or something in the commit message.

**Comment by trotzig** [comment]
Looks good to me so far.

File: lib/environments/meteorEnvironment.js (hunk 1 above)
Comment (lencioni) [inline lib/environments/meteorEnvironment.js]: Storybook has an `@` in its package name. https://github.com/kadirahq/react-storybook/blob/master/package.json#L2

Comment (rhettlivingston) [inline lib/environments/meteorEnvironment.js]: The `m` flag is definitely working as expected in the code and is needed. The [MDN documentation](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Regular_Expressions#Advanced_searching_with_flags) also includes it.

<package>@<version> is the syntax used in packages to pin a package to a version. I had no need for this particular version.

The best documentation I could find (probably because I trust code more than docs and didn't look extensively at the docs) is the [validatePackageName function](https://github.com/meteor/meteor/blob/devel/packages/package-version-parser/package-version-parser.js#L379) in Meteor's code. It looks for bad characters with the following statement.
```js
  var badChar = packageName.match(/[^a-z0-9:.\-]/);
```
So, "@" isn't allowed.

File: lib/environments/meteorEnvironment.js (hunk 2 above)
Comment (rhettlivingston) [inline lib/environments/meteorEnvironment.js]: I'm looking for the same. May have to just create it and place it here in a comment. "packages" is considered to be an internal file though most users maintain it by hand instead of using the provided `meteor add`, `meteor remove` and `meteor list` commands. It implicitly supports such usage by allowing comments (prefixed with "#") and blank lines.

File: lib/environments/meteorEnvironment.js (hunk 1 above)
Comment (trotzig) [inline lib/environments/meteorEnvironment.js]: My mistake about `m`. I was thinking of the dotall modifier (`s`).

**Comment by rhettlivingston** [comment]
@trotzig Thanks for pushing me to dig out the .meteor/packages syntax. It forced me to enough thought to realize that the code I had was including leading whitespace in the package name instead of stripping it. Comments have been added and corrections made.

**Comment by rhettlivingston** [comment]
While working to create the meteorPackageDependencies function and the meteorNamedExports function I'm noting what is going to become a performance issue.

I'm testing using `importjs fix` and these routines are being called multiple times, on an order related to the number of unique unresolved variables.

This won't matter for the more common `importjs word` or `importjs goto` but it seems like we should seek to avoid it in the case of `importjs fix`.

**Comment by rhettlivingston** [comment]
AFAIK, this is ready to go.

**Comment by trotzig** [comment]
Looks good to me. Thanks!

**Comment by trotzig** [comment]
I didn't look to closely at the performance issue. I see two ways to resolve that: 
- Move the call to `config.get('packageDependencies');` to an earlier stage in the import, and pass the array along to consumers (currently `findJsModulesFor` and `ImportStatements`). 
- Cache the results internally in `meteorEnvironment.js`. We can use `meteorPackagesPath` + last modification time of that file as the cache key. 

I would prefer the first option as it will allow other environments (including the default) to also benefit from not being called multiple times.

**Comment by rhettlivingston** [comment]
The first one sounds best to me for now. But, ultimately, I'm wondering if we'll eventually have to do the second.

My concern with the second is primarily complexity. The namedExports are much more important in this discussion than the packageDependencies. Detecting packageDependency changes that should force a rebuild of the cached value is easy. Detecting namedExport changes involves watching a lot more files. I think that in even a simple Meteor application, it is in the 100s.

So, for now, I think we should punt on building a cache mechanism that can work across multiple import-js operations and simply seek to reduce the actual invocations of the packageDependencies and namedExports functions to once per import-js operation.

But, we will likely evolve.

Note that the Meteor packageDependencies execution is currently a bit naive. It is going to evolve to be a bit module specific. In the Meteor environment, users frequently edit both their app modules and modules within their local packages in the same editing session. Apps and packages, even local packages located within the app directory structure, have different packageDependencies. The packageDependencies implementation I just submitted handles app  dependencies on packages. I've got to add the functionality for detecting Meteor package dependencies on packages.

Basically, I think I'm going to have to start ignoring the current working directory and determining the context of the module import-js is working with. This will involve walking up the directory tree. If I hit a directory containing a Meteor "package.js" file before I hit a directory containing ".meteor/packages", then the module is in a package and will need its packageDependencies determined differently. And, if I find it is under a "node_modules" directory with a package.json located above it, it's a third situation with a third means of determining package dependencies. All three situations could happen in a Meteor project.

**Comment by lencioni** [comment]
There might be another option. We could do a debounced memoization of `config.get('packageDependencies')` so that any call within a threshold returns the cached value.

**Comment by rhettlivingston** [comment]
>There might be another option. We could do a debounced memoization of config.get('packageDependencies') so that any call within a threshold returns the cached value.

I don't know enough about your daemon and batch operations to say for sure, but that sounds like you might be thinking of using the cache from one file to the next.

packageDependencies is definitely going to become a pathToCurrentFile dependent operation (within days hopefully) though there may be many files that get the same answer returned.

Any caching operation that spans files would thus have to use option 2 above where the code within the Meteor environment has the possibility of recognizing the current file's context and returning the particular cached answer for that context.

**Comment by trotzig** [comment]
I think it might be worth implementing time-consuming stuff in a background thread and accept the fact that changes aren't reflected immediately. Doing so probably involves leaning on the Watchman integration a lot more. We can generalize `WatchmanFileCache.js` so that we can hook in arbitrary jobs when files are changed.
